### PR TITLE
Interface should always start in first person WL 21423

### PR DESCRIPTION
**Pull request by Armads**
### To test prevention of Mirror mode:
1. Open Interface.
2. If camera is not in mirror mode, set camera to mirror mode.
3. Close and reopen Interface.
4. Camera should begin in first-person mode instead of mirror mode.

### To test saving/reloading of third-person mode:
1. Open Interface. Note whether the camera begins in first-person or third-person mode. Camera should not be in mirror mode (i.e. already pass the above test).
2. If camera began in first-person, switch to third-person. If camera began in third-person, switch to first-person.
3. Close and reopen Interface.
4. Camera should remain in the mode it was set to in step 2, and not revert to the mode noted in step 1.

### And to test that camera does not change modes without user action:
1. Open Interface. Note whether the camera begins in first-person or third-person mode. Camera should not be in mirror mode.
2. Close and reopen Interface.
3. Camera should remain in the mode it began in in step 1.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Can one of the admins verify this patch?

**Comment by themelissabrown** [comment]
Ok to test

**Comment by themelissabrown** [comment]
@Armads please provide a test plan for this

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1409/HighFidelity-Beta-PR10884-80f2a5c99d20e187d975792bdfbc3a035fd82de6.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1409/HighFidelity-Beta-PR10884-80f2a5c99d20e187d975792bdfbc3a035fd82de6.dmg

**Comment by Armads** [comment]
To test prevention of Mirror mode:
1. Open Interface.
2. If camera is not in mirror mode, set camera to mirror mode.
3. Close and reopen Interface.
4. Camera should begin in first-person mode instead of mirror mode.

To test saving/reloading of third-person mode:
1. Open Interface. Note whether the camera begins in first-person or third-person mode. Camera should not be in mirror mode (i.e. already pass the above test).
2. If camera began in first-person, switch to third-person. If camera began in third-person, switch to first-person.
3. Close and reopen Interface.
4. Camera should remain in the mode it was set to in step 2, and not revert to the mode noted in step 1.

And to test that camera does not change modes without user action:
1. Open Interface. Note whether the camera begins in first-person or third-person mode. Camera should not be in mirror mode.
2. Close and reopen Interface.
3. Camera should remain in the mode it began in in step 1.

Adding the above text to PR description.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1443/HighFidelity-Beta-PR10884-00b9a72481878b5dfef87cafe407a85c0cccbc14.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1443/HighFidelity-Beta-PR10884-00b9a72481878b5dfef87cafe407a85c0cccbc14.dmg

**Comment by CainFoool** [comment]
Testing now.

**Comment by themelissabrown** [comment]
@Armads are you able to see and rectify what's failing on the build? Or do I need to get you more information?

**Comment by Armads** [comment]
@themelissabrown no, I don't see any errors on my end. How is the code being pulled and built?

**Comment by Atlante45** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1518/HighFidelity-Beta-PR10884-01225f9f512aacaf46005a85c13a60d3c52c8250.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1518/HighFidelity-Beta-PR10884-01225f9f512aacaf46005a85c13a60d3c52c8250.dmg

**Comment by tapalisa** [comment]
My understanding is that this is only desired behavior on HMD. I don't think desktop users will appreciate this.

File: libraries/ui/src/ui/Menu.cpp [hunk 1]
```diff
@@ -71,15 +71,20 @@ void Menu::scanMenuBar(settingsAction modifySetting) {
...
 
-//    settings.beginGroup(menu.title());
-    foreach (QAction* action, menu.actions()) {
+    //    settings.beginGroup(menu.title());
+    foreach(QAction* action, menu.actions()) {
         if (action->menu()) {
             scanMenu(*action->menu(), modifySetting, settings);
-        } else if (action->isCheckable()) {
-            modifySetting(settings, *action);
+        }
```
Comment (ZappoMan) [inline libraries/ui/src/ui/Menu.cpp]: coding standard

File: libraries/ui/src/ui/Menu.cpp [hunk 2]
```diff
@@ -71,15 +71,20 @@ void Menu::scanMenuBar(settingsAction modifySetting) {
 void Menu::scanMenu(QMenu& menu, settingsAction modifySetting, Settings& settings) {
     groups.push_back(menu.title());
 
-//    settings.beginGroup(menu.title());
-    foreach (QAction* action, menu.actions()) {
+    //    settings.beginGroup(menu.title());
```
Comment (ZappoMan) [inline libraries/ui/src/ui/Menu.cpp]: delete dead code

File: libraries/ui/src/ui/Menu.cpp [hunk 3]
```diff
@@ -71,15 +71,20 @@ void Menu::scanMenuBar(settingsAction modifySetting) {
...
+        else if (action->isCheckable()) {
+            auto shouldSaveItr = _shouldSaveActionSetting.find(action);
+            if (shouldSaveItr != _shouldSaveActionSetting.end()
+                && *shouldSaveItr) {
+                modifySetting(settings, *action);
+            }
         }
     }
-//    settings.endGroup();
+    //    settings.endGroup();
```
Comment (ZappoMan) [inline libraries/ui/src/ui/Menu.cpp]: delete dead code

File: interface/src/Menu.cpp [hunk 4]
```diff
@@ -233,8 +233,8 @@ Menu::Menu() {
 
     // View > Mirror
     cameraModeGroup->addAction(addCheckableActionToQMenuAndActionHash(viewMenu,
-        MenuOption::FullscreenMirror, 0,
-        false, qApp, SLOT(cameraMenuChanged())));
+        MenuOption::FullscreenMirror, 0, // QML Qt::Key_H,
```
Comment (ZappoMan) [inline interface/src/Menu.cpp]: delete dead code

File: interface/src/Menu.cpp [hunk 5]
```diff
@@ -223,7 +223,7 @@ Menu::Menu() {
 
     // View > First Person
     cameraModeGroup->addAction(addCheckableActionToQMenuAndActionHash(viewMenu,
-        MenuOption::FirstPerson, 0,
+        MenuOption::FirstPerson, 0, // QML Qt:: Key_P
```
Comment (ZappoMan) [inline interface/src/Menu.cpp]: delete dead code

**Comment by ZappoMan** [comment]
@themelissabrown - Can you please double check that this is really the behavior we want. In particular, it seems like some users will be surprised to see that they now *always* have to explicitly switch into third person to use the app that way. The wording of the original WL job implies this might be more specifically related to "first run"

**Comment by themelissabrown** [comment]
@ZappoMan this is a request by @PhilipRosedale. I believe it should apply to HMD only as per email convo, but that isn't specified in the specs I have.

**Comment by themelissabrown** [comment]
@PhilipRosedale please test and clarify final functionality

**Comment by ZappoMan** [comment]
@themelissabrown - this code as written definitely impacts all users, not just HMD users. If the goal is to always start HMD users in first person, the code/approach would be different,

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1596/HighFidelity-Beta-PR10884-e3c0f19fa38c5877dfd168cf9085b7dda6acccb7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1596/HighFidelity-Beta-PR10884-e3c0f19fa38c5877dfd168cf9085b7dda6acccb7.exe

**Comment by themelissabrown** [comment]
@ZappoMan can you review his changes and I'll get clarification on where we're going from here?

**Comment by ZappoMan** [comment]
This latest PR is busted - it's adding back unrelated code that was recently removed. It looks like a bad merge/rebase.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1655/HighFidelity-Beta-PR10884-5f144b279af649e9ca7a2d2440425fd7a455b4da.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1655/HighFidelity-Beta-PR10884-5f144b279af649e9ca7a2d2440425fd7a455b4da.exe

**Comment by PhilipRosedale** [comment]
@themelissabrown @ZappoMan I agree that we should restrict this behavior to affect only HMD users.  What I want is for there to be no way for the HMD user to easily enter a mode that is confusing or impaired.  So what I would suggest is that on switching into HMD, we always switch into 1P.

**Comment by PhilipRosedale** [comment]
if a desktop user exits and restarts in 3P, they should stay in 3P

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1679/HighFidelity-Beta-PR10884-99fd52f6e8272438964af9e19aec85da14941055.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1679/HighFidelity-Beta-PR10884-99fd52f6e8272438964af9e19aec85da14941055.exe

**Comment by Armads** [comment]
The application already saves the user's preference if they switch to third-person explicitly; the next time they start the application they will still be in third-person and not first-person. This is noted in the testing procedure (specifically the second and third test cases). The only camera mode that the application will never start in is Mirrored.

Does this address what you're referring to or do you mean something else? Will investigate the HMD-only behavior but right now I think the desktop behavior is already as expected.

**Comment by tapalisa** [comment]
One use case this is intended to prevent someone from being in third person in HMD when they were just in third person in desktop. It seems like the right way to do this would be to save your preferences per mode, the way we do in the Audio menu. 

Either way, please update the test plan to reflect the conversation in this thread.

**Comment by zfox23** [comment]
I'm concerned about the interaction between this PR and the advanced Independent or Entity camera modes.

As this PR stands, quitting Interface after selecting Independent or Entity camera mode then restarting Interface results in the user being in Third Person Mode. This feels against the spirit of the PR.

To change this (if desired (it is of my opinion that it is)), I *think* that all you'd have to do is change `interface/src/Menu.cpp:243` and `interface/src/Menu.cpp:249` to include the same extra code you've already added to `interface/src/Menu.cpp:237`.

**Comment by themelissabrown** [comment]
OK, I'm going to have a QA tester re-test this in connection with the functionality in https://github.com/highfidelity/hifi/pull/10898 and specfically with the Independent or Entity camera modes to determine whether we merge it or not. Will email Alisa, Philip and Brad for clarification on specific functionality so we can determine a clear pass or fail.

**Comment by anshuman64** [comment]
I will let @ZappoMan make the final call, but here's what I would like to see.

In a user's FIRST SESSION (clean install):
- If they start in desktop mode (ie. no HMD plugged in), start them in 3P. Currently, many users are confused "what they are" because they cannot see their body. 
- If they start in HMD mode (ie. there is an HMD plugged in), start them in 1P.

For returning sessions:
In Desktop mode:
- **Quit in 1P, reenter in 1P**
- **Quit in 3P, reenter in 3P**
- **Quit in Mirror, reenter in 3P.** Mirror mode is very rarely a mode for default operations; I must have been doing something pretty particular during my last session to exit in mirror mode. I wouldn't want to start up in mirror mode again.
- **Quit in Independent, reenter in 3P.** I would be super confused if I quit Interface in independent mode a week ago and reopened Interface not seeing my avatar.
- **Quit in Entity, reenter in 3P.** Same explanation as independent mode.

In HMD:
- Quit in 1P, reenter in 1P.
- Quit in 3P, reenter in 1P--UNLESS "Enter First Person Mode in HMD Mode" option is OFF, in which case reenter in 3P
- Quit in Mirror, reenter in 1P
- Quit in Independent, reenter in 1P
- Quit in Entity, reenter in 1P

How does that sound?

**Comment by ZappoMan** [comment]
@anshuman64 - that sounds like a good spec to me.

**Comment by themelissabrown** [comment]
new PR 11110

**Pull request closed without merging**